### PR TITLE
replace setup.py integration with a warning

### LINKED ISSUE #330: Integration with `setup.py test` should not require `virtualenv` to be installed
- Bitbucket: https://bitbucket.org/hpk42/tox/issue/330
- Originally reported by: @bittner
- Originally created at: 2016-05-06T01:47:57.577

In my Python projects I want to separate the requirements only needed for testing cleanly from the actual project.

The beautiful thing of tox is I can specify all those additional requirements in the tox configuration (tox.ini), and with the [integration in setup.py](http://testrun.org/tox/latest/example/basic.html#integration-with-setuptools-distribute-test-commands) it's possible to not even have tox installed to execute tests.

Actually, this is the theory, because unfortunately, running tests with `python setup.py test` still requires one dependency to be installed beforehand: `virtualenv`

## Current Behavior

``` bash
$ python setup.py test
running test
Searching for tox
Reading https://pypi.python.org/simple/tox/
Best match: tox 2.3.1
Downloading https://pypi.python.org/packages/46/39/e15a857fda1852da1485bc88ac4268dbcef037ab526e1ac21accf2a5c24c/tox-2.3.1.tar.gz#md5=9371b3d3e25c03751a0372e19602dfb9
Processing tox-2.3.1.tar.gz
Writing /tmp/easy_install-2zqkvsk9/tox-2.3.1/setup.cfg
Running tox-2.3.1/setup.py -q bdist_egg --dist-dir /tmp/easy_install-2zqkvsk9/tox-2.3.1/egg-dist-tmp-f202aq90
...
Installed /home/user/repos/my-project/.eggs/tox-2.3.1-py3.4.egg
Searching for pluggy<0.4.0,>=0.3.0
Reading https://pypi.python.org/simple/pluggy/
Best match: pluggy 0.3.1
Downloading https://pypi.python.org/packages/1b/a9/6f5f80b75a8d84d21a8a13486fe26a2da9f043f93b464b2e3928be256dc4/pluggy-0.3.1.tar.gz#md5=ecdd791e309f60668b66fec97c2ee7db
Processing pluggy-0.3.1.tar.gz
Writing /tmp/easy_install-p0d1rfyw/pluggy-0.3.1/setup.cfg
Running pluggy-0.3.1/setup.py -q bdist_egg --dist-dir /tmp/easy_install-p0d1rfyw/pluggy-0.3.1/egg-dist-tmp-gsv46avp
...
Installed /home/user/repos/my-project/.eggs/pluggy-0.3.1-py3.4.egg
Searching for py>=1.4.17
Reading https://pypi.python.org/simple/py/
Best match: py 1.4.31
Downloading https://pypi.python.org/packages/f4/9a/8dfda23f36600dd701c6722316ba8a3ab4b990261f83e7d3ffc6dfedf7ef/py-1.4.31.tar.gz#md5=5d2c63c56dc3f2115ec35c066ecd582b
Processing py-1.4.31.tar.gz
Writing /tmp/easy_install-ax6qj5s2/py-1.4.31/setup.cfg
...
Installed /home/user/repos/my-project/.eggs/py-1.4.31-py3.4.egg
Searching for virtualenv>=1.11.2
Reading https://pypi.python.org/simple/virtualenv/
Best match: virtualenv 15.0.1
Downloading https://pypi.python.org/packages/c8/82/7c1eb879dea5725fae239070b48187de74a8eb06b63d9087cd0a60436353/virtualenv-15.0.1.tar.gz#md5=28d76a0d9cbd5dc42046dd14e76a6ecc
Processing virtualenv-15.0.1.tar.gz
Writing /tmp/easy_install-qkl88xj8/virtualenv-15.0.1/setup.cfg
...
Installed /home/user/repos/my-project/.eggs/virtualenv-15.0.1-py3.4.egg
running egg_info
creating my_project.egg-info
writing my_project.egg-info/PKG-INFO
...
writing manifest file 'my_project.egg-info/SOURCES.txt'
running build_ext
GLOB sdist-make: /home/user/repos/my-project/setup.py
flake8 create: /home/user/repos/my-project/.tox/flake8
ERROR: invocation failed (exit code 1), logfile: /home/user/repos/my-project/.tox/flake8/log/flake8-0.log
ERROR: actionid: flake8
msg: getenv
cmdargs: ['/home/user/.virtualenvs/my-project/bin/python', '-m', 'virtualenv', '--python', '/home/user/.virtualenvs/my-project/bin/python3.4', 'flake8']
env: {...}

/home/user/.virtualenvs/my-project/bin/python: No module named virtualenv

ERROR: InvocationError: /home/user/.virtualenvs/my-project/bin/python -m virtualenv --python /home/user/.virtualenvs/my-project/bin/python3.4 flake8 (see /home/user/repos/my-project/.tox/flake8/log/flake8-0.log)
_____________________________________________________________________________________ summary ______________________________________________________________________________________
ERROR:   flake8: InvocationError: /home/user/.virtualenvs/my-project/bin/python -m virtualenv --python /home/user/.virtualenvs/my-project/bin/python3.4 flake8 (see /home/user/repos/my-project/.tox/flake8/log/flake8-0.log)
```
When I run `pip install virtualenv` the same testrun runs like a charm.

Is this a problem of tox, or an issue of setuptools?

## Expected Behavior

After having set up tox with [setuptools integration](http://testrun.org/tox/latest/example/basic.html#integration-with-setuptools-distribute-test-commands)
it should be possible to run `python setup.py test` to execute tests driven by tox
without having had to install any requirements (such as `virtualenv` or `tox`) manually beforehand.

## References

This problem is also documented in [issue #245](https://bitbucket.org/hpk42/tox/issues/245/can-not-install-sure-dependency-in-tox#comment-26884414).

### LINKED ISSUE #375: remove documentation for setup.py integration
followup after #330 

integrating tox into setup.py deliberately breaks any standard packaging procedure
by suddenly elevating the testing from the *current* interpreter to *many* interpreters

this is a fundamental conceptual break that should not be inflicted on down-streams in any case

### PR DESCRIPTION
**Pull request by RonnyPfannschmidt**
fixes #375
fixes #330

### CODE REVIEW

File: doc/example/basic.txt [hunk 1]
```diff
@@ -237,47 +237,14 @@ If you wish to disable this feature, you can pass the command line option
...
-using the ``--tox-args`` or ``-a`` command-line options. For example::
-
-    python setup.py test -a "-epy27"
-
-is equivalent to running ``tox -epy27``.
+.. warning::
+
+  Integrating tox into ``setup.py`` is strongly discouraged as 
+  it breaks standard packaging approaches as used by many downstream distributions
+  which expect ``setup.py test`` to to test the current interpreter
```
Comment (bittner) [inline doc/example/basic.txt]: @RonnyPfannschmidt Can you please quote the "standard packaging approaches" and "many downstream distributions"? I don't know about any.

The argument sounds strongly biased. We should care about transparency.

Cc @hpk42

Comment (FRidh) [inline doc/example/basic.txt]: @bittner I agree it sounds biased and should therefore be rephrased. I don't think anyone has a good overview of how downstream distributions deal with this or similar situations. Perhaps we should document this somewhere? But where? And who will keep this up to date, as such information will always lag on the current practices of the distributions. 

[PEP 426](https://www.python.org/dev/peps/pep-0426) mentions the following:
```
python setup.py test : run the distribution's test suite in place given an sdist, source archive or VCS checkout
```
The distribution means here the package.

Neither this PEP, nor others (as far as I know), mention what the test suite can and cannot cover. Therefore, using `tox` like that is supposedly fine.

However, as a Python on Nix packager that would very much annoy me and I would argue that
>  `python setup.py test` is for testing against the current env, and `tox` is for testing against
multiple envs.

https://github.com/tox-dev/tox/issues/330#issuecomment-251644538

Comment (RonnyPfannschmidt) [inline doc/example/basic.txt]: @bittner the recommended approach is *not* calling tox from setup.py - it is basically outlined in the rest of the documentation

**Comment by hpk42** [comment]
In subsequent commits i revised the text.  See

https://github.com/tox-dev/tox/blob/master/doc/example/basic.txt#L237

The basic argument is that ``setup.py test`` should invoke the eventual test runner, not tox which creates multiple envs and does downloads from the web for it -- i doubt distros are happy about such test behaviour and think several maintainers confirm.

**Pull request merged**